### PR TITLE
Add test for byos integration

### PR DESCRIPTION
**Pull request by lukeseawalker**
Test does:
* get launch templates data and start a compute node
* test event handler execution and environment
* check artifacts are downloaded
* check artifacts created by head event are shared to compute
* check substack outputs are fetched
* check users are created and imds is set based on config
* check instance type data is fetched
* check cluster configuration is fetched
* check computes are terminated on cluster deletion

Tested with config:
```
{%- import 'common.jinja2' as common -%}
---
test-suites:
  byos:
    test_byos.py::test_byos:
      dimensions:
        - regions: ["eu-west-1"]
          instances: {{ common.INSTANCES_DEFAULT_X86 }}
          oss: ["alinux2"]
          schedulers: ["byos"]
```

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/tests/byos/test_byos.py [hunk 1]
```diff
@@ -9,19 +9,172 @@
...
 
-    cluster_config = pcluster_config_reader()
+    # Create bucket and upload resources
+    bucket_name = s3_bucket_factory()
+    bucket = boto3.resource("s3", region_name=region).Bucket(bucket_name)
+    for file in ["byos_infra.cfn.yaml", "artifact"]:
+        bucket.upload_file(str(test_datadir / file), f"byos/{file}")
+    # Create cluster
+    byos_user_list = deepcopy(BYOS_USERS_LIST)
+    cluster_config = pcluster_config_reader(bucket=bucket_name, user1=byos_user_list.pop(), user2=byos_user_list.pop())
```
Comment (demartinofra) [inline tests/integration-tests/tests/byos/test_byos.py:45]: nit: if you want to get rid of the copy you can just assign byos_user_list[0] and byos_user_list[1]

File: tests/integration-tests/tests/byos/test_byos.py [hunk 2]
```diff
@@ -9,19 +9,172 @@
...
+    return launch_templates_list
+
+
+def _start_compute_node(region, cluster, command_executor):
+    """Start and wait compute node to finish setup"""
+    # Get one launch template for compute node
+    lt = _get_launch_templates(command_executor)[0]
+    run_instance_command = (
+        f"aws ec2 run-instances --region {region}"
+        f" --instance-initiated-shutdown-behavior terminate"
```
Comment (demartinofra) [inline tests/integration-tests/tests/byos/test_byos.py:83]: isn't the shutdown behavior already baked into the LT we provide?

File: tests/integration-tests/tests/byos/test_byos.py [hunk 3]
```diff
@@ -9,19 +9,172 @@
...
+    compute_node_private_ip = compute_node.get("privateIpAddress")
+    compute_byos_log_output = command_executor.run_remote_command(
+        f"ssh -q {compute_node_private_ip} cat {BYOS_LOG_PATH}"
+    ).stdout
+    for event in ["ComputeInit", "ComputeConfigure", "ComputeFinalize"]:
+        assert_that(compute_byos_log_output).contains(f"[{event}] - INFO: {event} executed")
+        _test_event_handler_environment(event, compute_byos_log_output)
+
+
+def _test_event_handler_environment(event, log_output):
```
Comment (demartinofra) [inline tests/integration-tests/tests/byos/test_byos.py:136]: shall we assert the actual value of these variables to make sure we set them correctly?

File: tests/integration-tests/tests/byos/test_byos.py [hunk 4]
```diff
@@ -9,19 +9,172 @@
...
+    assert_that(home_listing).contains("artifact")
+
+
+def _test_subtack_outputs(command_executor):
+    """Test substack output is fetched by head node"""
+    head_byos_substack_outputs = command_executor.run_remote_command(f"cat {BYOS_SUBSTACK_OUTPUTS_PATH}").stdout
+    assert_that(head_byos_substack_outputs).contains('"TestOutput":"TestValue"')
+
+
+def _test_users(command_executor, compute_node):
```
Comment (demartinofra) [inline tests/integration-tests/tests/byos/test_byos.py:175]: we'll need to test imds access

File: tests/integration-tests/tests/byos/test_byos.py [hunk 5]
```diff
@@ -9,19 +9,172 @@
...
+LAUNCH_TEMPLATES_CONFIG_PATH = "/opt/parallelcluster/shared/launch_templates_config.json"
+BYOS_SUBSTACK_OUTPUTS_PATH = "/opt/parallelcluster/shared/byos_substack_outputs.json"
+BYOS_LOG_PATH = "/var/log/parallelcluster/byos-plugin.log"
+BYOS_HOME = "/home/byos"
+BYOS_USERS_LIST = ["user1", "byosUser"]
+
 
 @pytest.mark.usefixtures("instance", "scheduler", "os")
-def test_byos(region, pcluster_config_reader, clusters_factory):
+def test_byos(region, pcluster_config_reader, s3_bucket_factory, clusters_factory, test_datadir):
```
[the remark this answers is not in the document]
Comment (demartinofra) [reply tests/integration-tests/tests/byos/test_byos.py:34]: Other things we might want to test (they can come as incremental PRs):
* writing a file to the shared byos directories from the head node makes it available to event handlers in the compute?
* sudo privileges of the byos user
* is there some CW test we can reuse to validate logs in a BYOS cluster?

**Review by demartinofra — COMMENTED** [review]
LGTM. Dropped a few comments that could also be addressed incrementally in separate PRs

File: tests/integration-tests/tests/byos/test_byos.py (hunk 2 above)
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos.py:83]: it was not, added in https://github.com/aws/aws-parallelcluster/pull/3446

File: tests/integration-tests/tests/byos/test_byos.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos.py]: > need to add this to the tests config file
already there

> writing a file to the shared byos directories from the head node makes it available to event handlers in the compute?
DONE 

> sudo privileges of the byos user
> is there some CW test we can reuse to validate logs in a BYOS cluster?
left for future imporvements

File: tests/integration-tests/tests/byos/test_byos.py (hunk 1 above)
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos.py]: done

File: tests/integration-tests/tests/byos/test_byos.py (hunk 4 above)
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos.py:219]: added

File: tests/integration-tests/tests/byos/test_byos.py (hunk 3 above)
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos.py]: added

File: tests/integration-tests/tests/byos/test_byos/test_byos/pcluster.config.yaml [hunk 6]
```diff
@@ -14,11 +14,58 @@ Scheduling:
...
+            Command: env && echo "HeadClusterUpdate executed"
+        HeadComputeFleetStart:
+          ExecuteCommand:
+            Command: env && echo "HeadComputeFleetStart executed"
+        HeadComputeFleetStop:
+          ExecuteCommand:
+            Command: env && echo "HeadComputeFleetStop executed"
+      PluginResources:
+        ClusterSharedArtifacts:
+          - Source: https://github.com/aws/aws-parallelcluster/archive/refs/heads/develop.zip
```
[the remark this answers is not in the document]
Comment (lukeseawalker) [reply tests/integration-tests/tests/byos/test_byos/test_byos/pcluster.config.yaml]: done

**Review by demartinofra — APPROVED** [review]

**Review by demartinofra — APPROVED** [review]

**Pull request merged**